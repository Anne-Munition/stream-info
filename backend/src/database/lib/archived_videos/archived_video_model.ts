import { Document, Schema, model } from 'mongoose';

const schema = new Schema({
  videoId: String,
  createdAt: Date,
  length: Number,
  locked: Boolean,
});

export interface ArchiveVideoDoc extends Document {
  videoId: string;
  createdAt: string;
  length: number;
  locked: boolean;
}

export interface ArchiveVideoBulkUpdate {
  updateOne: {
    filter: { videoId: string };
    update: {
      videoId: string;
      createdAt: string;
      length: number;
    };
    upsert: boolean;
  };
}

export default model<ArchiveVideoDoc>('archived_videos', schema);
