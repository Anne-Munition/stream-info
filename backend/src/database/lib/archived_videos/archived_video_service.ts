import ArchiveVideo, {
  ArchiveVideoBulkUpdate,
  ArchiveVideoDoc,
} from './archived_video_model';

async function getLocked(): Promise<ArchiveVideoDoc[]> {
  return ArchiveVideo.find({ locked: true });
}

function createBulkOperation(
  video: TwitchVideo,
  duration: number,
): ArchiveVideoBulkUpdate {
  return {
    updateOne: {
      filter: { videoId: video.id },
      update: {
        videoId: video.id,
        createdAt: video.created_at,
        length: duration,
      },
      upsert: true,
    },
  };
}

async function bulkUpdate(ops: ArchiveVideoBulkUpdate[]): Promise<void> {
  await ArchiveVideo.bulkWrite(ops);
}

async function getInRange(
  start: string,
  end: string,
): Promise<ArchiveVideoDoc[]> {
  return ArchiveVideo.find({
    createdAt: { $gte: start, $lte: end },
  });
}

export default {
  getLocked,
  createBulkOperation,
  bulkUpdate,
  getInRange,
};
