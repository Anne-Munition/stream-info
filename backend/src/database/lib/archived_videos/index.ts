import ArchiveVideoService from './archived_video_service';

export default ArchiveVideoService;
