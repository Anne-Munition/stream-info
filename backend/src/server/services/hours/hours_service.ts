import dayjs from 'dayjs';
import timezone from 'dayjs/plugin/timezone';
import utc from 'dayjs/plugin/utc';
import moment, { Moment } from 'moment-timezone';
import ArchiveVideoService from '../../../database/lib/archived_videos';

dayjs.extend(utc);
dayjs.extend(timezone);

export default async (): Promise<{
  thisQuarter: number;
  lastQuarter: number;
}> => {
  const { startOfQuarter, startOfLastQuarter, endOfQuarter, endOfLastQuarter } =
    determineTimeRage();
  return {
    thisQuarter: await getHours(startOfQuarter, endOfQuarter),
    lastQuarter: await getHours(startOfLastQuarter, endOfLastQuarter),
  };
};

export function determineTimeRage() {
  const now = moment().tz('America/Los_Angeles');
  const startOfQuarter = qMonths(now.startOf('month'));
  const endOfQuarter = moment(startOfQuarter).add(2, 'months').endOf('month');
  const startOfLastQuarter = moment(startOfQuarter)
    .subtract(3, 'months')
    .startOf('month');
  const endOfLastQuarter = moment(endOfQuarter)
    .subtract(3, 'months')
    .endOf('month');
  return {
    startOfLastQuarter,
    endOfLastQuarter,
    startOfQuarter,
    endOfQuarter,
  };
}

export function qMonths(date: any) {
  const month = date.month() + 1;
  let sub = 0;
  while ((month - sub) % 3 !== 0) {
    sub++;
  }
  return date.subtract(sub, 'month');
}

export async function getHours(start: Moment, end: Moment): Promise<number> {
  const videos = await ArchiveVideoService.getInRange(
    start.toISOString(),
    end.toISOString(),
  );
  const sum = videos.reduce((prev, next) => {
    return prev + next.length;
  }, 0);
  return Math.round(sum / 60 / 60);
}
