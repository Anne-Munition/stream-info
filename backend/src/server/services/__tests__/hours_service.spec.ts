import dayjs from 'dayjs';
import timezone from 'dayjs/plugin/timezone';
import utc from 'dayjs/plugin/utc';
import MockDate from 'mockdate';
import moment from 'moment-timezone';
import ArchiveVideoService from '../../../database/lib/archived_videos';
import { ArchiveVideoDoc } from '../../../database/lib/archived_videos/archived_video_model';
import { determineTimeRage, getHours, qMonths } from '../hours/hours_service';

dayjs.extend(utc);
dayjs.extend(timezone);

describe('hoursStreamed module', () => {
  describe('determineTimeRage method', () => {
    const dates = {
      startOfLastQ4: moment.tz('2019-09-01 00:00:00', 'America/Los_Angeles').toLocaleString(),
      endOfLastQ4: moment.tz('2019-11-30 23:59:59.999', 'America/Los_Angeles').toLocaleString(),
      startOfQ1: moment.tz('2019-12-01 00:00:00', 'America/Los_Angeles').toLocaleString(),
      middleOfQ1: moment.tz('2020-01-15 00:00:00', 'America/Los_Angeles').toLocaleString(),
      endOfQ1: moment.tz('2020-02-29 23:59:59.999', 'America/Los_Angeles').toLocaleString(),
      startOfQ2: moment.tz('2020-03-01 00:00:00', 'America/Los_Angeles').toLocaleString(),
      middleOfQ2: moment.tz('2020-04-15 00:00:00', 'America/Los_Angeles').toLocaleString(),
      endOfQ2: moment.tz('2020-05-31 23:59:59.999', 'America/Los_Angeles').toLocaleString(),
      startOfQ3: moment.tz('2020-06-01 00:00:00', 'America/Los_Angeles').toLocaleString(),
      middleOfQ3: moment.tz('2020-07-15 00:00:00', 'America/Los_Angeles').toLocaleString(),
      endOfQ3: moment.tz('2020-08-31 23:59:59.999', 'America/Los_Angeles').toLocaleString(),
      startOfQ4: moment.tz('2020-09-01 00:00:00', 'America/Los_Angeles').toLocaleString(),
      middleOfQ4: moment.tz('2020-10-15 00:00:00', 'America/Los_Angeles').toLocaleString(),
      endOfQ4: moment.tz('2020-11-30 23:59:59.999', 'America/Los_Angeles').toLocaleString(),
      startOfNextQ1: moment.tz('2020-12-01 00:00:00', 'America/Los_Angeles').toLocaleString(),
      middleOfNextQ1: moment.tz('2021-01-15 00:00:00', 'America/Los_Angeles').toLocaleString(),
      endOfNextQ1: moment.tz('2021-02-28 23:59:59.999', 'America/Los_Angeles').toLocaleString(),
    };

    test('dates matches snapshot', () => {
      expect(dates).toMatchSnapshot();
    });

    testThing(
      'Q1',
      [dates.startOfQ1, dates.middleOfQ1, dates.endOfQ1],
      [dates.startOfLastQ4, dates.endOfLastQ4, dates.startOfQ1, dates.endOfQ1],
    );

    testThing(
      'Q2',
      [dates.startOfQ2, dates.middleOfQ2, dates.endOfQ2],
      [dates.startOfQ1, dates.endOfQ1, dates.startOfQ2, dates.endOfQ2],
    );

    testThing(
      'Q3',
      [dates.startOfQ3, dates.middleOfQ3, dates.endOfQ3],
      [dates.startOfQ2, dates.endOfQ2, dates.startOfQ3, dates.endOfQ3],
    );

    testThing(
      'Q4',
      [dates.startOfQ4, dates.middleOfQ4, dates.endOfQ4],
      [dates.startOfQ3, dates.endOfQ3, dates.startOfQ4, dates.endOfQ4],
    );

    testThing(
      'Next Q1',
      [dates.startOfNextQ1, dates.middleOfNextQ1, dates.endOfNextQ1],
      [dates.startOfQ4, dates.endOfQ4, dates.startOfNextQ1, dates.endOfNextQ1],
    );
  });

  function testThing(name: string, array: string[], expectedArray: string[]) {
    describe(name, () => {
      test.each(array)('%s', (date) => {
        MockDate.set(date);
        const { startOfLastQuarter, endOfLastQuarter, startOfQuarter, endOfQuarter } =
          determineTimeRage();
        expect(startOfLastQuarter.toLocaleString()).toBe(expectedArray[0]);
        expect(endOfLastQuarter.toLocaleString()).toBe(expectedArray[1]);
        expect(startOfQuarter.toLocaleString()).toBe(expectedArray[2]);
        expect(endOfQuarter.toLocaleString()).toBe(expectedArray[3]);
      });
    });
  }

  describe('getHours method', () => {
    it('returns a rounded number of hours', async () => {
      const mongoSpy = jest.spyOn(ArchiveVideoService, 'getInRange').mockImplementation(() => {
        return Promise.resolve([
          {
            length: 18000, // 5 hours
          },
          {
            length: 10800, // 3 hours
          },
          {
            length: 9360, // 2.6 Hours
          },
        ] as ArchiveVideoDoc[]);
      });
      const hours = await getHours(
        moment('2020-01-01T00:00:00.000Z'),
        moment('2020-01-10T00:00:00.000Z'),
      );

      expect(mongoSpy).toHaveBeenCalledWith('2020-01-01T00:00:00.000Z', '2020-01-10T00:00:00.000Z');
      expect(hours).toBe(11);
    });
  });

  describe('qMonths method', () => {
    test('December', () => {
      const actual = qMonths(dayjs('2019-12-01'));
      expect(actual.month() + 1).toBe(12);
    });

    test('January', () => {
      const actual = qMonths(dayjs('2020-01-01'));
      expect(actual.month() + 1).toBe(12);
    });

    test('February', () => {
      const actual = qMonths(dayjs('2020-02-01'));
      expect(actual.month() + 1).toBe(12);
    });

    test('March', () => {
      const actual = qMonths(dayjs('2020-03-01'));
      expect(actual.month() + 1).toBe(3);
    });

    test('April', () => {
      const actual = qMonths(dayjs('2020-04-01'));
      expect(actual.month() + 1).toBe(3);
    });

    test('May', () => {
      const actual = qMonths(dayjs('2020-05-01'));
      expect(actual.month() + 1).toBe(3);
    });

    test('June', () => {
      const actual = qMonths(dayjs('2020-06-01'));
      expect(actual.month() + 1).toBe(6);
    });

    test('July', () => {
      const actual = qMonths(dayjs('2020-07-01'));
      expect(actual.month() + 1).toBe(6);
    });

    test('August', () => {
      const actual = qMonths(dayjs('2020-08-01'));
      expect(actual.month() + 1).toBe(6);
    });

    test('September', () => {
      const actual = qMonths(dayjs('2020-09-01'));
      expect(actual.month() + 1).toBe(9);
    });

    test('October', () => {
      const actual = qMonths(dayjs('2020-10-01'));
      expect(actual.month() + 1).toBe(9);
    });

    test('November', () => {
      const actual = qMonths(dayjs('2020-11-01'));
      expect(actual.month() + 1).toBe(9);
    });
  });
});
