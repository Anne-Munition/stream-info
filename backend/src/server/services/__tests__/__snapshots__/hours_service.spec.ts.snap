// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`hoursStreamed module determineTimeRage method dates matches snapshot 1`] = `
Object {
  "endOfLastQ4": "Sat Nov 30 2019 23:59:59 GMT-0800",
  "endOfNextQ1": "Sun Feb 28 2021 23:59:59 GMT-0800",
  "endOfQ1": "Sat Feb 29 2020 23:59:59 GMT-0800",
  "endOfQ2": "Sun May 31 2020 23:59:59 GMT-0700",
  "endOfQ3": "Mon Aug 31 2020 23:59:59 GMT-0700",
  "endOfQ4": "Mon Nov 30 2020 23:59:59 GMT-0800",
  "middleOfNextQ1": "Fri Jan 15 2021 00:00:00 GMT-0800",
  "middleOfQ1": "Wed Jan 15 2020 00:00:00 GMT-0800",
  "middleOfQ2": "Wed Apr 15 2020 00:00:00 GMT-0700",
  "middleOfQ3": "Wed Jul 15 2020 00:00:00 GMT-0700",
  "middleOfQ4": "Thu Oct 15 2020 00:00:00 GMT-0700",
  "startOfLastQ4": "Sun Sep 01 2019 00:00:00 GMT-0700",
  "startOfNextQ1": "Tue Dec 01 2020 00:00:00 GMT-0800",
  "startOfQ1": "Sun Dec 01 2019 00:00:00 GMT-0800",
  "startOfQ2": "Sun Mar 01 2020 00:00:00 GMT-0800",
  "startOfQ3": "Mon Jun 01 2020 00:00:00 GMT-0700",
  "startOfQ4": "Tue Sep 01 2020 00:00:00 GMT-0700",
}
`;
